### PR TITLE
Receiver mode GUI

### PR DESCRIPTION
**Pull request by micahflee**
As you can see, I've written a **lot** more code. This implements the GUI portion of receiver mode, which included tackling many thorny issues. There are still a few more things that need to be done before this feature is complete, including:

* In ReceiveMode, make `Uploads.open_folder()` actually open folders with specific file selected in Windows and macOS
* In `ReceiveMode.timeout_finished_should_stop_server()`, wait for current uploads to finish before stopping the server, but still prevent new uploads from starting
* Related, if I start the server in receive mode, start uploading large files, then stop the server before they've uploaded, the upload history window still shows the in-progress upload. If I then start the server again, it crashes
* Make sure `Settings.build_default_downloads_dir()` works in Windows
* Update `install/onionshare.nsi` to make the Windows installer/uninstaller use all the new filenames
* More thoroughly test the toggle button for Uploads/Downloads, making sure it actually scrolls to the bottom at the appropriate times, etc.
* Improve the language of strings, based on feedback from @glenn-sorrentino
* Change receive mode web interface, based on feedback from @glenn-sorrentino
* In settings dialog, persistent address is a common setting, not just for share mode
* In Upload History window, really long filenames should be word-wrapped
* If an upload fails to finish, it should be marked as failed in the upload history
* You should be able to clear upload history
* Prevent 0 byte, filename-less files (this happened when I made a public receive mode for some reason)
* Update "uploads complete" and "uploads in progress" numbers, and change the mouseover text
* I think this is it...?

Now, here's an explanation of what's new. First, there are a few new settings:

screenshot_2018-05-20_15-36-22

You can choose which folder to save files to.

"Receive mode can be stopped by the sending" adds an "I'm Finished Uploading" to the web interface, and anyone who sends you files can click it to close the server. You enable this if you want to allow other people to close your server. (And disable it if you want a long-standing server, or a public one.)

"Receive mode is open to the public" completely removes the slug, and also gets rid of the 404 rate-limiting. So instead of the URL being `http://[something].onion/[slug]`, it's just `http://[something].onion`. Normally with a slug, after 20 404 errors the server automatically shuts down to prevent guessing the slug, but this option disables that. (Otherwise, anyone could DoS the server by making 20 404 requests.)

Now, here's what the CLI interface looks like:

```
user@dev ~/code/onionshare (receiver-mode-gui) $ ./dev_scripts/onionshare --receive
OnionShare 1.3 | https://onionshare.org/
Configuring onion service on port 17641.
Starting ephemeral Tor onion service and awaiting publication
Settings saved to /home/user/.config/onionshare/onionshare.json
Preparing files to share.
 * Running on http://127.0.0.1:17641/ (Press CTRL+C to quit)

Files people send you will appear in this folder: /home/user/OnionShare

Warning: Receive mode lets someone else upload files to your computer. Some files can hack your computer if you open them! Only open files from people you trust, or if you know what you're doing.

Give this address to the people sending you files:
http://ebpuq7tlgwb27nmy.onion

Press Ctrl+C to stop the server
127.0.0.1 - - [20/May/2018 15:41:44] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [20/May/2018 15:41:45] "GET /static/img/favicon.ico HTTP/1.1" 200 -
127.0.0.1 - - [20/May/2018 15:41:46] "GET /static/css/style.css HTTP/1.1" 200 -
127.0.0.1 - - [20/May/2018 15:41:47] "GET /static/img/logo_large.png HTTP/1.1" 200 -
127.0.0.1 - - [20/May/2018 15:41:47] "GET /static/img/logo.png HTTP/1.1" 200 -
May 20, 03:42PM: Upload of total size 4.8 MiB is starting
=> 4.6 MiB         OnionShare.UX.Spec.pdf
=> 1.4 KiB         README.md
=> 176.1 KiB       onionshare-1.3-1.noarch.rpm
Received file: /home/user/OnionShare/OnionShare.UX.Spec.pdf
Received file: /home/user/OnionShare/README.md
Received file: /home/user/OnionShare/onionshare-1.3-1.noarch.rpm
127.0.0.1 - - [20/May/2018 15:42:44] "POST /upload HTTP/1.1" 302 -
127.0.0.1 - - [20/May/2018 15:42:46] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [20/May/2018 15:42:52] "POST /close HTTP/1.1" 200 -
```

Now when you open OnionShare in GUI mode, there are two tabs, Share and Receive:

screenshot_2018-05-20_15-43-39
screenshot_2018-05-20_15-43-46

When you click "Start Receive Mode", the onion service starts just like in share mode, and eventually the server starts:

screenshot_2018-05-20_15-45-21
screenshot_2018-05-20_15-46-01

Here's what the web interface looks like for receive mode (though it will change some, Glenn gave some good suggestions like moving the location and color of the close button):

screenshot_2018-05-20_15-47-52

For the person uploading the files, there's no progress bar, it's just a straight up HTTP POST request. I'm not sure if we can improve this without requiring javascript in the browser (or maybe we can use javascript if it's available and gracefully degrade if it's not?). But for now, this is what the web interface looks like after you successfully upload files:

screenshot_2018-05-20_15-49-31

The upload history window in the GUI interface (that you access by clicking that little button) looks different though. For each HTTP POST request (which might include multiple file uploads) it shows one progress bar, and you can see the status of each file as it arrives. For example:

screenshot_2018-05-20_15-56-51

Also notice in that screenshot that the files in the top of the upload history end in `-2`, like `README-2.md`, even though the uploaded file was `README.md` -- this is because there was already a file called `README.md` in the uploads folder, so it intelligently renamed it.

There are likely many small details I've left out. But for now, I just have a small to-do list left to do before it's time to review this PR. And then, we can finally start working on a handful of other much smaller issues for the 2.0 release.

### CODE REVIEW

**Comment by b-meson** [comment]
I think the GUI uploaded needs a nicer look than the current red text. Maybe something that mimics the behavior of python flask's flash with a green background.

File: dev_scripts/onionshare [hunk 1]
```diff
@@ -3,7 +3,7 @@
 """
 OnionShare | https://onionshare.org/
 
-Copyright (C) 2014-2018 Micah Lee <micah@micahflee.com>
+Copyright (C) 2014-2018 Micah Lee <micah@micahflee.com>>>>>>>> develop
```
Comment (mig5) [inline dev_scripts/onionshare]: whoops!

Comment (micahflee) [reply dev_scripts/onionshare]: heh oops

**Comment by micahflee** [comment]
Ok I just opened #748 to track further receive mode work. As soon as you feel like this PR is in good enough shape to merge, I say go ahead and merge.

Then we can update #731 to merge into develop instead.

**Comment by micahflee** [comment]
Hey @mig5 do you want to give one final review to make sure this is in good shape, then merge it into develop?

We can change #752 to merge into develop instead of receive-mode-gui.

**Comment by mig5** [comment]
I see the same shutdown_slug 404 in receiver-mode-gui branch that I observed in another branch recently

```
[Sep 18 2018 10:04:48 AM] Web.check_slug_candidate: slug_candidate=iurpwgkhpmoceyhjwevaj6zzl4, slug_compare=iurpwgkhpmoceyhjwevaj6zzl4
127.0.0.1 - - [18/Sep/2018 10:04:48] "GET /iurpwgkhpmoceyhjwevaj6zzl4/shutdown HTTP/1.1" 404 -
```

I suspect public_mode broke it.

Another observation: the 'uploads in progress/completed' icons aren't right - a finished upload stays in the 'upload in progress' green icon, and the green 'tick' (upload completed) icon never fires. I think you raised this in another ticket though, so it's perhaps not worth blocking this merge for this.

I want to spend some time looking at that shutdown slug issue though, because it's causing the Flask web server to stay running. Fortunately, the .onion is no longer available because that cleanup still completes.

**Pull request merged**